### PR TITLE
Migrate modes into the parsetree

### PR DESCRIPTION
**Pull request by freemagma**
This PR migrates the mode syntax from attributes to parsetree.

An older version of this PR is at https://github.com/freemagma/flambda-backend/pull/1, where most conversations are now resolved.

### CODE REVIEW

File: ocaml/parsing/printast.ml [hunk 1]
```diff
@@ -135,17 +135,41 @@ let typevars ppf vs =
...
+(* let mode i ppf mode =
+ *   string_loc i ppf
+ *     (Location.map (fun (Mode x) -> x) mode)
+ *
+ * let modes ?(print_even_if_empty = false) i ppf modes =
+ *   match modes with
+ *   | [] when not print_even_if_empty -> ()
+ *   | _ ->
+ *     line i ppf "modes\n";
+ *     list i mode ppf modes *)
```
Comment (riaqn) [inline ocaml/parsing/printast.ml]: We don't need to keep this (can always add back if needed in the future)

File: ocaml/typing/printtyped.ml [hunk 2]
```diff
@@ -383,9 +383,8 @@ and expression_extra i ppf x attrs =
       line i ppf "Texp_constraint\n";
       attributes i ppf attrs;
       option i core_type ppf ct;
-      list i string ppf (
-        List.map (fun {txt = Parsetree.Mode x; _} -> x) modes
-      )
+      ignore modes;
+      (* CR cgunn: print modes *)
```
Comment (riaqn) [inline ocaml/typing/printtyped.ml:387]: I'm not sure what `modes` is ignored?

File: ocaml/typing/typemode.ml [hunk 3]
```diff
@@ -52,6 +52,33 @@ let transl_mode_annots modes =
   in
   loop Alloc.Const.Option.none modes
 
+let untransl_mode_annots ~loc (modes : Mode.Alloc.Const.Option.t) =
+  let print_to_string print a =
```
Comment (riaqn) [inline ocaml/typing/typemode.ml:56]: We can probably use `Format.asprintf` for this.

File: ocaml/typing/typemode.ml [hunk 4]
```diff
@@ -52,6 +52,33 @@ let transl_mode_annots modes =
...
+    let buf = Buffer.create 32 in
+    let fmt = Format.formatter_of_buffer buf in
+    print fmt a;
+    Format.pp_print_flush fmt ();
+    Buffer.contents buf
+  in
+  let areality =
+    Option.map (print_to_string Mode.Locality.Const.print) modes.areality
```
Comment (riaqn) [inline ocaml/typing/typemode.ml:64]: we can probably push the `Option.map` thing into `print_to_string`.

File: ocaml/typing/typecore.ml [hunk 5]
```diff
@@ -5928,8 +5928,10 @@ and type_expect_
         exp_type = ty';
         exp_attributes = arg.exp_attributes;
         exp_env = env;
-        exp_extra = (Texp_constraint (Some extra_cty, modes), loc, sexp.pexp_attributes)
-                    :: arg.exp_extra;
+        exp_extra =
+          (Texp_constraint (Some extra_cty, Typemode.transl_mode_annots modes),
```
Comment (riaqn) [inline ocaml/typing/typecore.ml:5932]: `transl_mode_annots` is called twice - we should factor that out. Not just for performance reasons, but also it's more "correct"

File: ocaml/testsuite/tests/parsetree/modes_ast_mapper.ml [hunk 6]
```diff
@@ -2,34 +2,42 @@
...
+  }
 
 let test mapper s =
   let p = Lexing.from_string s |> Parse.implementation in
   ignore (mapper.Ast_mapper.structure mapper p);
-  Format.printf "%a\n" (Printast.structure 0) p;
   Format.printf "------------------------------\n"
 
-(* CR zqian: add [modalities] to mapper so the following [bar hello] can be
-   printed *)
```
Comment (riaqn) [inline ocaml/testsuite/tests/parsetree/modes_ast_mapper.ml:25]: Thanks for taking care of this!

File: ocaml/testsuite/tests/typing-modes/modes.ml [hunk 7]
```diff
@@ -5,25 +5,25 @@
 (* Let bindings *)
 let local_ foo : string @@ unique = "hello"
 [%%expect{|
-Line 1, characters 4-43:
+Line 1, characters 36-43:
 1 | let local_ foo : string @@ unique = "hello"
-        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+                                        ^^^^^^^
```
Comment (riaqn) [inline ocaml/testsuite/tests/typing-modes/modes.ml:10]: I still think we should fix this - my reply is at https://github.com/freemagma/flambda-backend/pull/1#discussion_r1659058967

**Review by riaqn — COMMENTED** [review]
Several small things.

File: ocaml/parsing/printast.ml (hunk 1 above)
Comment (riaqn) [reply ocaml/parsing/printast.ml]: (This is removed by a later commit)

File: ocaml/typing/typecore.ml [hunk 8]
```diff
@@ -2794,15 +2776,21 @@ and type_pat_aux
...
+        let cty, ty, expected_ty' =
+          let type_modes = Typemode.transl_alloc_mode ms in
+          solve_Ppat_constraint ~refine tps loc env type_modes sty expected_ty
+        in
+        let p = type_pat ~alloc_mode tps category sp_constrained expected_ty' in
+        let extra = (Tpat_constraint cty, loc, sp_constrained.ppat_attributes) in
+        { p with pat_type = ty; pat_extra = extra::p.pat_extra }
+      | None ->
+        let p = type_pat ~alloc_mode tps category sp_constrained expected_ty in
+        p
```
Comment (riaqn) [inline ocaml/typing/typecore.ml:2792]: this can be simplified

File: ocaml/typing/printtyped.ml (hunk 2 above)
Comment (freemagma) [reply ocaml/typing/printtyped.ml]: Fixed

File: ocaml/testsuite/tests/typing-modes/modes.ml (hunk 7 above)
Comment (freemagma) [reply ocaml/testsuite/tests/typing-modes/modes.ml]: Fixed this. Now the modes are always a part of the pattern and expression locs that get created from `vb_pat_constraint` and `vb_expr_constraint`

File: ocaml/typing/typecore.ml [hunk 9]
```diff
@@ -5920,11 +5902,19 @@ and type_expect_
...
+      { exp with exp_loc = loc }
+  | Pexp_constraint (sarg, Some sty, modes) ->
+      let mode_annots = Typemode.transl_mode_annots modes in
+      let (ty, extra_cty) =
+        let alloc_mode =
+          Mode.Alloc.Const.Option.value mode_annots ~default:Mode.Alloc.Const.legacy
+        in
+        type_constraint env sty alloc_mode
       in
+      let expected_mode = type_expect_mode ~loc ~env ~modes expected_mode in
```
Comment (freemagma) [inline ocaml/typing/typecore.ml:5917]: This function calls `transl_mode_annots` again

File: ocaml/parsing/parsetree.mli [hunk 10]
```diff
@@ -1067,6 +1075,7 @@ and value_binding =
     pvb_pat: pattern;
     pvb_expr: expression;
     pvb_constraint: value_constraint option;
+    pvb_modes: modes;
```
[the remark this answers is not in the document]
Comment (riaqn) [reply ocaml/parsing/parsetree.mli:1136]: Resolved offline.

**Review by riaqn — APPROVED** [review]
LGTM (please address the new comment about `pprintast.ml`)

File: ocaml/parsing/pprintast.ml [hunk 11]
```diff
@@ -339,18 +340,37 @@ let maybe_atat_modalities f m =
...
+  | [] -> ()
+  | m -> pp f " %@%@ %a" modes m
+
+let split_out_legacy_modes =
+  List.partition (fun m ->
+    let Mode txt = m.txt in
+    match txt with
+    | "local" | "unique" | "once" -> true
+    | _ -> false
+  )
```
[the remark this answers is not in the document]
Comment (riaqn) [reply ocaml/parsing/pprintast.ml:366]: And please update `pprintast_unconditional.ml` as well:
```
  let modality_record  = parse module_expr
    "struct \
      type t = {global_ x : string @@ shared portable } \
     end"
  let modality_cstrarg = parse module_expr
    "struct \
      type t = Foo of global_ string @@ shared portable * global_ string @@ shared portable \
      type u = Foo : global_ string @@ shared portable * global_ string @@ shared portable-> u \
     end"
```

**Comment by goldfirere** [comment]
Why did this get closed? Is there another PR this work is continued on?

**Comment by ncik-roberts** [comment]
The PR was automatically closed as the result of me merging a PR. Charlie and I are figuring out how to re-open it.

**Comment by ncik-roberts** [comment]
Apparently I had perms to reopen it after Charlie undeleted the branch, but Charlie did not. Great! It's reopened.

**Pull request merged**